pkg/plugin/search_api.go

```diff
@@ -0,0 +1,174 @@
+	}
+	if queryParams != nil {
+		req.URL.RawQuery = queryParams.Encode()
+	}
+	backend.Logger.Debug("http GET", "URL", req.URL.String())
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("GET request failed: %v", err.Error())
+	}
+	defer res.Body.Close()
+	responseBody, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %v", err.Error())
+	}
+	if res.StatusCode != http.StatusOK {
+		// Try to parse the error from the response
+		if err := parseErrorFromResponse(responseBody); err != nil {
+			return nil, err
+		}
+		// Couldn't parse the error from the response, just return a generalized error
+		return nil, fmt.Errorf("request failed (%v): %v", res.StatusCode, string(responseBody))
+	}
+	return responseBody, nil
+}
+
+// Compose the full URL to an API resource
+func (api *SearchAPI) url(path string) string {
+	return fmt.Sprintf("%s%s", api.Settings.CriblOrgBaseUrl, path)
+}
+
+// Add the Authorization header to an http.Request, refreshing our cached authentication as needed
+func (api *SearchAPI) addAuthorization(req *http.Request) error {
+	err := api.refreshBearerTokenAsNeeded()
+	if err != nil {
+		return err
+	}
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", api.BearerToken.Token))
+	return nil
+}
+
+// Establish the cached bearer token, refreshing as needed.  This honors the expiration time
+// but applies a 30-second buffer to avoid cutting it too close.  Bearer tokens are typically
+// valid for many hours.
+func (api *SearchAPI) refreshBearerTokenAsNeeded() error {
+	if api.BearerToken != nil && api.BearerToken.ExpiresAt > (time.Now().UnixMilli()+30000) {
+		backend.Logger.Debug("Reusing cached bearer token", "ExpiresAt", api.BearerToken.ExpiresAt)
+		return nil // current token is still valid
+	}
+
+	backend.Logger.Debug("Refreshing bearer token")
+	var err error
+	if strings.HasSuffix(api.Settings.CriblOrgBaseUrl, ".cloud") { // i.e. foo.cribl.cloud or bar.cribl-staging.cloud
+		api.BearerToken, err = RefreshTokenViaOAuth(api.Settings.CriblOrgBaseUrl, api.Settings.ClientId, api.Settings.Secrets.ClientSecret)
+	} else {
+		api.BearerToken, err = RefreshTokenViaLocalAPI(api.Settings.CriblOrgBaseUrl, api.Settings.ClientId, api.Settings.Secrets.ClientSecret)
+	}
+	return err
+}
+
+// Try to parse an error from an API response.  Returns nil if for any reason we couldn't
+// parse the error, or if the response was an unexpected format.  We attempt to unpack the body
+// to whatever extent possible.  Normally we expect a JSON object with "status" and "message"
+// fields.  Often the "message" field itself is a serialized JavaScript Error object.  We make
+// an attempt to provide the most user-friendly representation of the error.
+func parseErrorFromResponse(body []byte) error {
+	var jsonBody map[string]interface{}
+	if err := json.Unmarshal(body, &jsonBody); err != nil {
```

Comment:
Should this case return an error?

Reply:
I specifically decided not to return `(error, error)` here.  I felt that would be royally confusing.  This is a function that tries to parse an error, and "failure" here is deliberately indicated by `nil`.

The equivalent of `return undefined` in TS.  This is one of those times when Go conventions can get in the way of clarity.

Reply:
BTW, in case it wasn't clear, if there's an error trying to unmarshal the JSON, it's not well-formed JSON, so we're returning `nil` to indicate that we couldn't parse an error from the JSON (that's not JSON).  :-)